ciao-controller/cnci.go

```diff
@@ -0,0 +1,397 @@
+func (c *CNCIManager) AddSubnet(subnet int) error {
+	c.cnciLock.Lock()
+
+	cnci, ok := c.subnets[subnet]
+	if ok {
+		// subnet already exists
+		c.cnciLock.Unlock()
+		return errors.New("Subnet already added")
+	}
+
+	cnci = &CNCI{
+		ctrl:      c.ctrl,
+		eventCh:   make(chan event),
+		stateLock: &sync.RWMutex{},
+		state:     starting,
+		subnet:    subnet,
+	}
+
+	c.subnets[subnet] = cnci
+
+	// send a launch command
+	// TBD: we are going to have to be able to tell the CNCI
+	// which subnet to manage eventually.
+	ID, err := c.launch()
+	if err != nil {
+		c.cnciLock.Unlock()
+		return err
+	}
+
+	cnci.ID = ID
+	c.cncis[ID] = cnci
+
+	c.cnciLock.Unlock()
+
+	// we release the lock before waiting because
+	// we need to be able to read the event channel.
+	return cnci.waitForEventTimeout(added, 1*time.Minute)
+}
+
+// RemoveSubnet is called when a subnet no longer is needed.
+// a cnci can be stopped.
+func (c *CNCIManager) RemoveSubnet(subnet int) error {
+	c.cnciLock.Lock()
+
+	cnci, ok := c.subnets[subnet]
+	if !ok {
+		// there is no cnci for this subnet
+		c.cnciLock.Unlock()
+		return errors.New("Subnet doesn't exist")
+	}
+
+	// TBD: this will change when we have multiple CNCIs per tenant
+	// we should really not be messing with this at all.
+	err := c.ctrl.ds.RemoveTenantCNCI(c.tenant)
+	if err != nil {
+		c.cnciLock.Unlock()
+		return err
+	}
+
+	err = cnci.stop()
+	if err != nil {
+		c.cnciLock.Unlock()
+		return err
+	}
+
+	c.cnciLock.Unlock()
+
+	err = cnci.waitForEventTimeout(removed, 1*time.Minute)
+	if err != nil {
+		return err
+	}
+
+	// remove subnet & cnci
+	c.cnciLock.Lock()
+	delete(c.cncis, cnci.ID)
+	delete(c.subnets, subnet)
+	c.cnciLock.Unlock()
+
+	return nil
+}
+
+// ConcentratorInstanceRemoved will move the CNCI back to the initial state
+// and send an event through the event channel.
+func (c *CNCIManager) ConcentratorInstanceRemoved(id string) error {
+	c.cnciLock.Lock()
+	defer c.cnciLock.Unlock()
+
+	cnci, ok := c.cncis[id]
+	if !ok {
+		return errors.New("No CNCI found")
+	}
+
+	cnci.ID = ""
+
+	cnci.transitionState(initial)
+
+	return nil
+}
+
+// ConcentratorInstanceAdded will move the CNCI into the active state
+// and send an event through the event channel.
+func (c *CNCIManager) ConcentratorInstanceAdded(id string) error {
+	c.cnciLock.Lock()
+	defer c.cnciLock.Unlock()
+
+	cnci, ok := c.cncis[id]
+	if !ok {
+		return errors.New("No CNCI found")
+	}
+
+	cnci.ID = id
+
+	cnci.transitionState(active)
+
+	return nil
+}
+
+// StartFailure will move the CNCI to the error state and
+// send an event through the event channel.
+func (c *CNCIManager) StartFailure(id string) error {
+	c.cnciLock.Lock()
+	defer c.cnciLock.Unlock()
+
+	cnci, ok := c.cncis[id]
+	if !ok {
+		return errors.New("No CNCI found")
+	}
+
+	cnci.ID = ""
+
+	cnci.transitionState(failed)
+
+	return nil
+}
+
+// WaitForActive will block until a cnci for a subnet is active
+func (c *CNCIManager) WaitForActive(subnet int) error {
```

Comment:
I think this function can go away if the same channel is used to signify that the initial cnci launch vs a later "wait"